oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,284 @@
+        val target = modelService.simByName[oldName]
+        if (newNameSimulation == null && target != null) {
+            modelService.renameSim(target, newName, oldName)
+            NodeNameChangeResponseDTO.newBuilder().setChanged(true).build()
+        } else {
+            NodeNameChangeResponseDTO.newBuilder().setChanged(false).build()
+        }
+    }
+
+    override fun register(request: RegistrationCommandDTO, responseObserver: StreamObserver<OASISQueryDTO>) {
+        if (request.name in modelService.simByName.keys) {
+            return
+        }
+        modelService.addNewSim(request.name!!, Simulation(emptyList(), emptyList(), "", responseObserver, Channel(1), Channel(1), Channel(1)))
+        modelService.updateStatusMessage("${request.name} registered")
+    }
+
+    override fun unregisterAll() {
+        throw NotImplementedException()
+    }
+
+    override fun startOptimization(request: StartOptimizationCommandDTO, responseObserver: StreamObserver<StartOptimizationResponseDTO>) = responseObserver.consume {
+        startOptimization(RandomNumberOptimizer())
+        StartOptimizationResponseDTO.newBuilder().setMessage("Started").setStarted(true).build()
+    }
+
+
+    override fun startOptimization(optimizer: Optimizer) {
+        GlobalScope.launch {
+            state = State.Running
+
+            for (simName in modelService.simByName.keys) {
+                syncConfigFor(simName)
+            }
+
+            modelService.updateStatusMessage("Evaluating...")
+
+            while (state == State.Running) {
+                var toolNumber = 1
+                for ((simName, sim) in modelService.simByName) {
+                    modelService.updateStatusMessage("Evaluating: $simName ($toolNumber/${modelService.simByName.keys.size})")
+                    val inputVector = optimizer.getInputs(sim.inputs)
+                    val message = OASISQueryDTO.newBuilder()
+                            .setEvaluationRequest(OASISQueryDTO.SimulationEvaluationRequest.newBuilder()
+                                    .setName(simName)
+                                    .putAllInputVector(inputVector)
+                            )
+                            .build()
+
+                    sim.input.onNext(message)
+                    val result = select<SimResult> {
+                        sim.output.onReceive { SimResult.Success(it.name, it.outputVectorMap) }
+                        sim.error.onReceive { SimResult.Failure(it.name, it.exception) }
+                        if (sim.timeOut != null) {
+                            onTimeout(sim.timeOut.toMillis()) {
+                                //Timed out
+                                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(simName)
+                                ).build()
+                                sim.input.onNext(message)
```

Comment:
Oh I see, this is where you try to wait for the cancellation.

Well it sure isn't idiomatic, but I suppose if it works safely, and I hope that since Roman tagged `onReceive(block: suspend (T) -> R)` that he knew that said `block` might suspend....

I've only ever thought of `block` as a pure adaptor. I would've written it as such and probably needed an extra data class (or perhalps expressed a null-result as cancellation)....

```kt
input.onNext(EvaluationRequest)
val result = select<SimResult?> {
  sim.output.onReceive { Success }
  sim.error.onReceive { Failure }
  onTimeout(duration) { null }
}

if(result == null){
  input.onNext(CancellationRequestBuilder.build())
  select {
    sim.output.onReceive { TimedOut(it.name) }
    sim.error.onReceive { TimeoutFailure(it) }
    //ahhhh, and you also dont use a timeout here, so failure to respond to a cancellation effectively halts the optimization.
  }
}
```

mine has less nesting I suppose? It also allows you to push functionality up stack, which I realize now is generally a good thing.